src/prove.rs

```diff
@@ -130,11 +130,11 @@ pub struct Prover8_56 {
 impl Prover8_56 {
     pub(crate) const NONCES_PER_AES: u32 = 16;
 
-    pub fn new<P: pow::Prover>(
+    pub fn new(
         challenge: &[u8; 32],
         nonces: Range<u32>,
         params: ProvingParams,
-        pow_prover: &P,
+        pow_prover: Arc<dyn pow::Prover>,
```

Comment:
the problem is in `service/src/service.rs`, I can't really figure out how to do this without boxing in the value:
```rust
                let threads = self.threads.clone();
                let stop = self.stop.clone();
                let progress = ProvingProgress::default();
                let pow_prover: dyn post::pow::Prover = match &self.remote_k2pow {
                    Some(uri) => post::pow::service::K2powService::new(uri.clone()), <- breaks compilation
                    None => post::pow::randomx::PoW::new(pow_flags).unwrap(),
                };
                let reporter = progress.clone();
                *proof_gen = ProofGenProcess::Running {
                    challenge,
                    handle: Some(std::thread::spawn(move || {
                        post::prove::generate_proof(
                            &datadir,
                            &challenge,
                            cfg,
                            nonces,
                            threads,
                            pow_flags,
                            stop,
                            reporter,
                            &pow_prover,
                        )
                    })),
                    progress,
                };

```

Reply:
It needs to be boxed there because the two branches of the `match` return a different type. This is perfectly fine IMHO:
```rust
let pow_prover: Box<dyn post::pow::Prover> = match &self.remote_k2pow {
    Some(uri) => Box::new(post::pow::service::K2powService::new(uri.clone())),
    None => Box::new(post::pow::randomx::PoW::new(pow_flags).unwrap()),
};
```

Reply:
Sure, but then how can `pow_prover` type be `pow_prover: &dyn pow::Prover` as you requested to change it? Shouldn't it be `pow_prover: Box<dyn pow::Prover>`?

Reply:
You can just take a reference `&pow_prover`. It will work because `Box` [implements](https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-Deref-for-Box%3CT,+A%3E) [`std::ops::Deref`](https://doc.rust-lang.org/std/ops/trait.Deref.html#) and because deref coercion. Specifically:

> If T implements Deref<Target = U>, and v is a value of type T, then:
> - Values of type &T are coerced to values of type &U